servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpConnector.java

```diff
@@ -209,4 +206,88 @@ private NoopNettyAddressResolverGroup() {
             return NOOP_ADDRESS_RESOLVER;
         }
     }
+
+    private static final class ConnectHandler<C extends ListenableAsyncCloseable> implements Consumer<Channel> {
+        private static final Logger LOGGER = LoggerFactory.getLogger(ConnectHandler.class);
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<ConnectHandler> terminatedUpdater =
+                newUpdater(ConnectHandler.class, "terminated");
+
+        private final DelayedCancellable futureCancellable = new DelayedCancellable();
+        private final DelayedCancellable flatMapCancellable = new DelayedCancellable();
+        private final Subscriber<? super C> target;
+        private final Function<Channel, Single<? extends C>> connectionFactory;
+
+        private volatile int terminated;
+
+        ConnectHandler(final Subscriber<? super C> target,
+                       final Function<Channel, Single<? extends C>> connectionFactory) {
+            this.target = target;
+            this.connectionFactory = connectionFactory;
+            target.onSubscribe(() -> {
+                try {
+                    futureCancellable.cancel();
+                } finally {
+                    flatMapCancellable.cancel();
+                }
+            });
+        }
+
+        @Override
+        public void accept(final Channel channel) {
+            toSource(connectionFactory.apply(channel)
+                    .subscribeShareContext())
+                    .subscribe(new Subscriber<C>() {
+                        @Override
+                        public void onSubscribe(final Cancellable cancellable) {
+                            flatMapCancellable.delayedCancellable(cancellable);
+                        }
+
+                        @Override
+                        public void onSuccess(@Nullable final C connection) {
+                            if (terminatedUpdater.compareAndSet(ConnectHandler.this, 0, 1)) {
+                                target.onSuccess(connection);
+                            } else {
+                                LOGGER.info("Connection {} created for a channel: {} but connect failed previously. " +
+                                                "Closing connection",
+                                        connection, channel);
```

Comment:
nit: This logging message will read a bit strange and may be confusing if `connection == null`.

Reply:
mmm ya but as we do not expect the connection to be `null` and I want to add a log if we are closing the connection, I think it is OK to not add more conditional to handle `null`. I will change the level to `debug` though as there isn't anything a user is expected to do here.